src/MDMTransaction.m

```diff
@@ -16,31 +16,108 @@
 
 #import "MDMTransaction.h"
 #import "MDMTransaction+Private.h"
+#import "MDMPerforming.h"
+#import "MDMPlan.h"
 
 @implementation MDMTransaction {
   NSMutableArray *_logs;
+  NSMutableSet *_namedPlans;
 }
 
 - (instancetype)init {
   self = [super init];
   if (self) {
     _logs = [NSMutableArray array];
+    _namedPlans = [NSMutableSet set];
   }
   return self;
 }
 
 - (void)addPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target {
-  MDMTransactionLog *log = [MDMTransactionLog new];
-  log.plans = @[ [plan copy] ];
-  log.target = target;
+  [self commonAddPlan:plan toTarget:target withName:nil];
+}
+
+- (void)addPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target withName:(NSString *)name {
+  [self removePlanNamed:name fromTarget:target];
+  [self commonAddPlan:plan toTarget:target withName:name];
+}
+
+- (void)commonAddPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target withName:(NSString *)name {
+  MDMTransactionLog *log = [self newLogWithPlan:plan target:target name:name];
+  if (name.length) {
+    Class performerClass = [plan performerClass];
+    id performer = [[performerClass alloc] initWithTarget:target];
+    if ([performer conformsToProtocol:@protocol(MDMNamedPlanPerforming)] &&
+        [performer respondsToSelector:@selector(addPlan:withName:)]) {
+      [performer addPlan:plan withName:name];
+    }
+    [_namedPlans addObject:log];
+  }
   [_logs addObject:log];
 }
 
+- (void)removePlanNamed:(nonnull NSString *)name fromTarget:(nonnull id)target {
+  MDMTransactionLog *log = [self newLogWithPlan:nil target:target name:name];
+  if ([_namedPlans containsObject:log]) {
+    // need to get the log back from _namedPlans so we can get at the plans
+    MDMTransactionLog *retrievedLog = [_namedPlans member:log];
+    if ([target isEqual:retrievedLog.target]) {
+      for (id<MDMPlan>plan in [retrievedLog plans]) {
+        Class performerClass = [plan performerClass];
+        id performer = [[performerClass alloc] initWithTarget:[retrievedLog target]];
+        if ([performer conformsToProtocol:@protocol(MDMNamedPlanPerforming)] &&
+            [performer respondsToSelector:@selector(removePlan:withName:)]) {
+          [performer removePlan:plan withName:name];
+        }
+      }
+    }
+    [_namedPlans removeObject:retrievedLog];
+    [_logs removeObject:retrievedLog];
```

Comment:
Ah! I see what was happening here now.

It may help to think about the transaction object as being completely unaware of the scheduler's internals. For example, the transaction object might be on another computer. When we "commit" the transaction to a scheduler on another computer we encode each of the recorded operations (add, addNamed, removeWithName) and send them over the wire. Then scheduler then takes action on each of these operations as though these methods were called directly on the scheduler.

The reason we funnel everything through the transaction object — rather than exposing APIs directly on the scheduler — is to ensure that we will architecturally move towards a place where tooling can interact with the runtime over a wire.